File: ios/LivtetJigsaw/Sources/LivtetJigsaw/Model/Manufacturer.swift
public final class Manufacturer {
    public var insertsGenerator: InsertsGenerator = InsertSequence.fixed
    public var metadata: [[String: Any]] = []
    public var headAnchor: Anchor?
    private var puzzleOptions: (pieceRadius: CGFloat, proximity: CGFloat) = (50, 1)
    private var puzzleWidth: Int = 5
    private var puzzleHeight: Int = 5

    public init() {}

    public func withMetadata(_ m: [[String: Any]]) -> Self { metadata = m; return self }
    public func withInsertsGenerator(_ g: @escaping InsertsGenerator) -> Self { insertsGenerator = g; return self }
    public func withHeadAt(_ anchor: Anchor) -> Self { headAnchor = anchor; return self }
    public func withStructure(_ opts: (pieceRadius: CGFloat, proximity: CGFloat)) -> Self { puzzleOptions = opts; return self }
    public func withDimensions(_ width: Int, _ height: Int) -> Self { puzzleWidth = width; puzzleHeight = height; return self }

    public func build() -> Puzzle {
        let puzzle = Puzzle(pieceRadius: puzzleOptions.pieceRadius, proximity: puzzleOptions.proximity)
        let positioner = Positioner(puzzle: puzzle, headAnchor: headAnchor)

        for y in 0..<puzzleHeight {
            let vertSeq = InsertSequence(insertsGenerator)
            let horizSeq = InsertSequence(insertsGenerator)

            for x in 0..<puzzleWidth {
                let left = horizSeq.previousComplement()
                let up = vertSeq.previousComplement()
                let right = horizSeq.current(puzzleWidth)
                let down = vertSeq.current(puzzleHeight)

                _ = horizSeq.next()
                _ = vertSeq.next()

                let piece = puzzle.newPiece(
                    structure: Structure(up: up, down: down, left: left, right: right)
                )
                piece.centerAround(positioner.naturalAnchor(x, y))
            }
        }

        annotateAll(puzzle.pieces)
        return puzzle
    }

    private func annotateAll(_ pieces: [Piece]) {
        for (i, piece) in pieces.enumerated() {
            let baseMeta = i < metadata.count ? metadata[i] : [:]
            var meta: [String: Any] = [:]
            baseMeta.forEach { meta[$0] = $1 }
            meta["id"] = meta["id"] as? String ?? String(i + 1)
            piece.annotate(meta)
        }
    }
}

private struct Positioner {
    let puzzle: Puzzle
    let offset: Vector

    init(puzzle: Puzzle, headAnchor: Anchor?) {
        self.puzzle = puzzle
        self.offset = headAnchor?.asVector() ?? puzzle.pieceDiameter
    }

    func naturalAnchor(_ x: Int, _ y: Int) -> Anchor {
        Anchor(
            x: CGFloat(x) * offset.x + offset.x,
            y: CGFloat(y) * offset.y + offset.y
        )
    }
}

File: android/jigsaw/src/main/kotlin/net/olamaelcu/livtet/jigsaw/model/Manufacturer.kt
package net.olamaelcu.livtet.jigsaw.model

class Manufacturer {
    var insertsGenerator: InsertsGenerator = twoAndTwoGenerate
    var metadata: List<Map<String, Any>> = emptyList()
    var headAnchor: Anchor? = null
    var width: Int = 4
    var height: Int = 4
    var pieceSize: Size = Size(Vector(50f, 50f))
    var proximity: Float = 10f

    fun withStructure(size: Size, proximity: Float) { pieceSize = size; this.proximity = proximity }
    fun withMetadata(m: List<Map<String, Any>>) { metadata = m }
    fun withInsertsGenerator(g: InsertsGenerator) { insertsGenerator = g }
    fun withHeadAt(a: Anchor) { headAnchor = a }
    fun withDimensions(w: Int, h: Int) { width = w; height = h }

    fun build(): Puzzle {
        val puzzle = Puzzle(pieceSize = pieceSize, proximity = proximity)
        val positioner = Positioner(puzzle, headAnchor)
        var verticalSequence = InsertSequence(insertsGenerator)

        for (y in 0 until height) {
            val horizontalSequence = InsertSequence(insertsGenerator)
            verticalSequence.next()
            for (x in 0 until width) {
                horizontalSequence.next()
                val piece = puzzle.newPiece(
                    left = horizontalSequence.previousComplement(),
                    up = verticalSequence.previousComplement(),
                    right = horizontalSequence.current(width),
                    down = verticalSequence.current(height),
                )
                piece.centerAround(positioner.naturalAnchor(x, y))
            }
        }
        annotateAll(puzzle.pieces)
        return puzzle
    }

    private fun annotateAll(pieces: List<Piece>) {
        pieces.forEachIndexed { index, piece ->
            val base = metadata.getOrNull(index) ?: emptyMap()
            val data = base.toMutableMap()
            if (!data.containsKey("id")) data["id"] = (index + 1).toString()
            piece.annotate(data)
        }
    }
}

internal class Positioner(private val puzzle: Puzzle, headAnchor: Anchor?) {
    private val offset: Vector = headAnchor?.asVector() ?: puzzle.pieceSize.radius

    fun naturalAnchor(x: Int, y: Int): Anchor =
        Anchor.anchor(
            x * puzzle.pieceSize.diameter.x + offset.x,
            y * puzzle.pieceSize.diameter.y + offset.y,
        )
}
